### PR TITLE
Create powersync persister

### PR DESCRIPTION
**Pull request by bndkt**
## Summary

Following on the heels of https://github.com/tinyplex/tinybase/pull/131, this is the PR to add the PowerSync persister to TinyBase. I'm submitting this in draft mode to give you a preview @jamesgpearce. Unfortunately I'm struggling to get the Jest unit tests working because the PowerSync SDK relies on Workers which are not available in Jest.

## How did you test this change?

Implemented the persister in a React Native demo app: https://github.com/bndkt/powerchat-tinybase/blob/tinybase/src/app/(app)/tinybase.tsx

Jest unit tests are failing, like described above.

### CODE REVIEW

File: src/persisters/persister-powersync.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  createSqlitePersister(
+    store,
+    configOrStoreTableName,
+    async (sql: string, args: any[] = []): Promise<IdObj<any>[]> =>
+      await promiseNew(async (resolve, reject) => {
+        const result = await powerSync.execute(sql, args);
+
+        result?.rows ? resolve(result.rows._array) : reject();
+      }),
+    (listener: UpdateListener): AbortController => {
```
Comment (jamesgpearce) [inline src/persisters/persister-powersync.ts:29]: 'AbortController' is a strange name for their listener handle!

File: src/types/docs/persisters/persister-powersync.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+{
+  /**
+   * The getDb method returns a reference to the database instance the Store is
+   * being persisted to.
+   * @returns A reference to the database instance.
+   * @example
+   * This example creates a Persister object against a newly-created Store and
+   * then gets the database instance back out again.
+   *
+   * ```js yolo
```
Comment (jamesgpearce) [inline src/types/docs/persisters/persister-powersync.js:29]: hahaha you found the 'yolo' trick :)

File: src/types/docs/persisters/persister-powersync.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
+/**
+ * The persister-powersync module of the TinyBase project lets you save and
+ * load Store data to and from a local SQLite database that is automaticaly
+ * synced using the PowerSync service.
+ * @see Persisting Data guide
+ * @packageDocumentation
+ * @module persister-powersync
+ * @since v4.6.9
```
Comment (jamesgpearce) [inline src/types/docs/persisters/persister-powersync.js:8]: We'll make this feature into the 4.7 release.

**Review by jamesgpearce — CHANGES_REQUESTED** [review]
Again, amazing what you've been able to figure out in this codebase :)

I'm tempted to say we don't need to worry too much about the tests. We don't explicitly test expo and electric either, because the environments are so hard to replicate.

But is there anyway to mock workers in Jest? eg jsdom-worker? I realize this may not be trivial, but since this is the only persister using the conflict flag, we may need to do something just to keep the coverage at 100%

File: src/persisters/persister-powersync.ts (hunk 1 above)
Comment (bndkt) [reply src/persisters/persister-powersync.ts:29]: Agreed, but they just use the standard https://developer.mozilla.org/en-US/docs/Web/API/AbortController

File: src/persisters/persister-powersync.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      const abortController = new AbortController();
+
+      (async () => {
+        for await (const update of powerSync.onChange({
```
[the remark this answers is not in the document]
Comment (bndkt) [reply src/persisters/persister-powersync.ts]: powerSync.onChange returns an async iterable, which I don't know that there is another way to loop over this. I've made this piece a bit cleaner and easier to read, but I think there's no way around the IIFE.

**Comment by bndkt** [comment]
Thanks for the feedback again @jamesgpearce! Regarding the test, I tried using jsdom-worker, tried mocks, but couldn't get it to run. I spoke to PowerSync, they're working on making it work with Jest themselves, so maybe in the future there will be a chance to add this. What I did for now is a bit unconventional, curious what you think. In the end PowerSync is just using a sqlite database locally and the sync part with their service wouldn't be part of the tests anyways. So I created a mock (https://github.com/bndkt/tinybase/blob/46dcf0e96fed009963294cfc2c366d606f206413/test/unit/persisters/sqlite.ts#L52-L118, happy to move this into a separate file) which uses a simple sqlite database (just like the sqlite persister) but mimics the PowerSync interface. Now I "only" get 14 failing tests which are related to granular saves and the difference in SQL statements. I'll have to spent a few brain cycles on granular saves and the missing ON CONFLICT capability in PowerSync, will post about that separately.

File: src/persisters/persister-powersync.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jamesgpearce) [reply src/persisters/persister-powersync.ts]: Well ok - but it will be the only `for` in the whole codebase :)

**Comment by jamesgpearce** [comment]
I'm hanging out here but take it out of draft when you want me to take a final look

**Comment by jamesgpearce** [comment]
Hey @bndkt is this ready? 😄  - do you want me to review and/or rebase this for you?

**Comment by bndkt** [comment]
Hey @jamesgpearce, thanks for checking in! I've poured a lot of time into fixing the tests now, but I'm afraid I need your help at this point. If you run the unit tests in this PR, you'll find 15 failing test. From my understanding, those can be put into two buckets:

- If you set useOnConflict to true for the PowerSync persister (exactly [here](https://github.com/bndkt/tinybase/blob/97ace7c8694f9f1ad872e60e75c16870bbaec9e5/src/persisters/persister-powersync.ts#L45)), there's only 3 tests that fail, all of which are related to autoLoad. Those must be due to a problem with my PowerSync [mock implementation](https://github.com/bndkt/tinybase/blob/97ace7c8694f9f1ad872e60e75c16870bbaec9e5/test/unit/persisters/sqlite.ts#L52-L123) but I can't figure out what the problem is.
- The remaining 12 failures are test for partial persists. I tried to implement partial changes without using ON CONFLICT (basically following the "good but tedious" option from [here](https://stackoverflow.com/a/4330694) but end up with a lot of tedious code and am not really sure about the value of this. Do you think it would be feasible to just don't write partial updates for this persister? From what I understand this would only mean rewriting some columns that are unchanged, or am I missing something?

**Comment by jamesgpearce** [comment]
OK I'll take a look! I found some broken tests in the v( beta branch and this might be similar. Leave it with me.

**Comment by jamesgpearce** [comment]
@bndkt please take a look at https://github.com/tinyplex/tinybase/pull/133/commits/51f1a2989e0dbafa87135cd40aec6f28566c44d4 - I simplified and isolated a simple (JSON-based) test that fails. The second write to the table (that has `"c1":2`) does not fire a change and so the test fails.

I read [here](https://docs.powersync.com/usage/installation/client-side-setup/instantiate-powersync-database) that there should not be more than one instance per database file. Maybe that's why you had the factory pattern, but it still doesn't seem to be a singleton. However, that didn't seem to fix it.

How well does this mock work do you think?

**Comment by jamesgpearce** [comment]
@bndkt a gentle ping to see if you have any thoughts on this mock issue

**Comment by bndkt** [comment]
@jamesgpearce Thanks for isolating this! I'm looking into it now, and see how to improve the mock in order to get the tests to pass.

**Comment by bndkt** [comment]
@jamesgpearce in order to debug the mock function, I've temporarily enabled "on conflict" usage for the persister, so that it should behave exactly like the SQLite3 persister. I now get 2 failing tests. I've added comments to those and will tag you. They both seem to be timing-related, do you have any idea why the timing could be off here?

File: test/unit/persisters/persister-sqlite-tabular.test.ts [hunk 5]
```diff
@@ -1476,6 +1476,7 @@ describe.each(Object.entries(VARIANTS))(
         expect(store2.getContent()).toEqual([{t1: {r1: {c1: 1}}}, {v1: 1}]);
       });
 
+      // FAILS
```
Comment (bndkt) [inline test/unit/persisters/persister-sqlite-tabular.test.ts:1479]: @jamesgpearce 

when I duplicate all the `await pause(autoLoadPause);` calls in this test (to await the autoLoadPause two times), this test succeeds.

Screenshot 2024-03-26 at 11 55 44 AM

File: test/unit/persisters/persisters.test.ts [hunk 6]
```diff
@@ -707,6 +713,7 @@ describe.each([
     expect(persister.getStats()).toEqual({loads: 1, saves: 0});
   });
 
+  // FAILS
```
Comment (bndkt) [inline test/unit/persisters/persisters.test.ts:716]: @jamesgpearce

File: test/unit/persisters/persister-sqlite-tabular.test.ts (hunk 5 above)
Comment (jamesgpearce) [reply test/unit/persisters/persister-sqlite-tabular.test.ts:1479]: I can't repro success by adding in the second pause. It still doesn't work. No 'change' event seems to fire from the mock, so the autoloading doesn't know to pick up a change.

**Comment by bndkt** [comment]
Oh wow, thanks for discovering this @jamesgpearce! I'll have a final look now and then set it to ready for review.

**Comment by bndkt** [comment]
Good to go @jamesgpearce!

**Comment by jamesgpearce** [comment]
Alright! Let's do any final polish in `main`.

**Comment by jamesgpearce** [comment]
Thanks so much for your work on this!

**Pull request merged**